File: problems/049/049.f90
PROGRAM F049

    ! Copyright 2021 Melwyn Francis Carlo
    ! FILE REFERENCE: http://www.naturalnumbers.org/primes.html

    IMPLICIT NONE

    INTEGER, PARAMETER :: N = 10

    INTEGER :: PRIME_NUM
    CHARACTER (LEN=50), PARAMETER :: FILENAME = "problems/003/PrimeNumbers_Upto_1000000"

    INTEGER, DIMENSION (1200, N)  :: PRIMES_NUMS = 0
    INTEGER, DIMENSION (1200)     :: PRIMES_DIGITS_OCCURRENCES = 0

    CHARACTER (LEN=N)                   :: PRIME_DIGITS  = ACHAR(0)
    CHARACTER (LEN=N), DIMENSION (1200) :: PRIMES_DIGITS = ACHAR(0)

    LOGICAL :: DUPLICATE_FOUND, SEQUENCE_FOUND

    INTEGER :: I, J, K, L, M, TEMP_VAL

    OPEN (2, FILE = FILENAME, STATUS = 'OLD')

    I = 1
    DO

        READ(2, *, END=10) PRIME_NUM

        IF (PRIME_NUM < 1000) CYCLE

        IF (PRIME_NUM > 9999) CYCLE

        PRIME_DIGITS = "0000000000"

        TEMP_VAL = INT(PRIME_NUM / 1000) + 1
        PRIME_DIGITS(TEMP_VAL:TEMP_VAL) = "1"
        TEMP_VAL = MOD(INT(PRIME_NUM / 100), 10) + 1
        PRIME_DIGITS(TEMP_VAL:TEMP_VAL) = "1"
        TEMP_VAL = MOD(INT(PRIME_NUM /  10), 10) + 1
        PRIME_DIGITS(TEMP_VAL:TEMP_VAL) = "1"
        TEMP_VAL = MOD(PRIME_NUM, 10) + 1
        PRIME_DIGITS(TEMP_VAL:TEMP_VAL) = "1"

        DUPLICATE_FOUND = .FALSE.
        DO J = 1, I

            IF (PRIMES_DIGITS(J) == PRIME_DIGITS) THEN

                IF (PRIMES_NUMS(J, N) == 0) THEN

                    DO K = 2, N
                        IF (PRIMES_NUMS(J, K) == 0) THEN
                            PRIMES_NUMS(J, K) = PRIME_NUM
                            EXIT
                        END IF
                    END DO

                    PRIMES_DIGITS_OCCURRENCES(J) = PRIMES_DIGITS_OCCURRENCES(J) + 1

                END IF

                DUPLICATE_FOUND = .TRUE.
                EXIT

            END IF

        END DO

        IF (.NOT. DUPLICATE_FOUND) THEN
            PRIMES_DIGITS_OCCURRENCES(I) = 0
            PRIMES_NUMS(I, 1) = PRIME_NUM
            PRIMES_DIGITS(I)  = PRIME_DIGITS
            I = I + 1
        END IF

    END DO

    10 CLOSE(2)

    J = 1
    K = 1
    L = 1
    M = 1
    SEQUENCE_FOUND = .FALSE.

    DO J = 1, I

        IF (PRIMES_DIGITS_OCCURRENCES(J) < 3) CYCLE

        DO K = 1, (N - 2)

            IF (PRIMES_NUMS(J, K) == 0) EXIT

            IF (PRIMES_NUMS(J, K) == 1487) CYCLE

            DO L = (K + 1), (N - 1)

                IF (PRIMES_NUMS(J, L) == 0) EXIT

                DO M = (L + 1), N
                    IF (PRIMES_NUMS(J, M) == 0) EXIT

                    IF ((PRIMES_NUMS(J, M) - PRIMES_NUMS(J, L))  &
                    ==  (PRIMES_NUMS(J, L) - PRIMES_NUMS(J, K))) THEN
                        SEQUENCE_FOUND = .TRUE.
                        EXIT
                    END IF
                END DO

                IF (SEQUENCE_FOUND) EXIT

            END DO

            IF (SEQUENCE_FOUND) EXIT

        END DO

        IF (SEQUENCE_FOUND) EXIT

    END DO

    PRINT ('(3I0)'), PRIMES_NUMS(J, K), PRIMES_NUMS(J, L), PRIMES_NUMS(J, M)

END PROGRAM F049

File: problems/049/049.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Copyright 2021 Melwyn Francis Carlo */
/* File Reference: http://www.naturalnumbers.org/primes.html */

#define N 10

int main()
{
    FILE *fp;
    char prime_num[10];
    const char* FILE_NAME = "problems/003/PrimeNumbers_Upto_1000000";
    fp = fopen(FILE_NAME, "r");

    int  primes_nums[1200][N]            = { 0 };
    char primes_digits[1200][11]         = { 0 };
    int  primes_digits_occurrences[1200] = { 0 };

    int i = 0;
    while (fgets(prime_num, 10, fp) != NULL)
    {
        if (atoi(prime_num) < 1000)
            continue;

        if (atoi(prime_num) > 9999)
            continue;

        char prime_digits[11] = "0000000000";

        prime_digits[prime_num[0]-'0'] = '1';
        prime_digits[prime_num[1]-'0'] = '1';
        prime_digits[prime_num[2]-'0'] = '1';
        prime_digits[prime_num[3]-'0'] = '1';

        int duplicate_found = 0;
        for (int j = 0; j < i; j++)
        {
            if (strcmp(primes_digits[j], prime_digits) == 0)
            {
                if (primes_nums[j][N-1] == 0)
                {
                    for (int k = 1; k < N; k++)
                    {
                        if (primes_nums[j][k] == 0)
                        {
                            primes_nums[j][k] = atoi(prime_num);
                            break;
                        }
                    }

                    primes_digits_occurrences[j]++;
                }

                duplicate_found = 1;
                break;
            }
        }

        if (!duplicate_found)
        {
            primes_digits_occurrences[i] = 0;
            primes_nums[i][0] = atoi(prime_num);
            memcpy(&primes_digits[i++][0], &prime_digits[0], 10);
        }
    }

    int j = 0;
    int k = 0;
    int l = 0;
    int m = 0;
    int sequence_found = 0;
    for (j = 0; j < i; j++)
    {
        if (primes_digits_occurrences[j] < 3)
            continue;

        for (k = 0; k <= (N - 3); k++)
        {
            if (primes_nums[j][k] == 0)
                break;

            if (primes_nums[j][k] == 1487)
                continue;

            for (l = (k + 1); l <= (N - 2); l++)
            {
                if (primes_nums[j][l] == 0)
                    break;

                for (m = (l + 1); m < N; m++)
                {
                    if (primes_nums[j][m] == 0)
                        break;

                    if ((primes_nums[j][m] - primes_nums[j][l]) 
                    ==  (primes_nums[j][l] - primes_nums[j][k]))
                    {
                        sequence_found = 1;
                        break;
                    }
                }

                if (sequence_found)
                    break;
            }

            if (sequence_found)
                break;
        }

        if (sequence_found)
            break;
    }

    fclose(fp);

    printf("%d%d%d\n", primes_nums[j][k], primes_nums[j][l], primes_nums[j][m]);

    return 0;
}
